### ISSUE TITLE
Incorporating SDR objects into the NetworkAPI

### ISSUE THREAD

**Issue by dkeeney** [htm-community/nupic.cpp#230]
Before I propose how I want to incorporate the SDR class into NetworkAPI, I want to make sure everyone understands what an Array class does.    
    
The Array class (a subclass of ArrayBase) is the container to hold various types of data.  The Array class is passed from Region to Region via the Links.  Each region has a set of Input buffers and Output buffers for different types of data and they are defined in the region's 'Spec'.  This spec defines the type (and sometimes the size) of the data that it expects as input and provides as output. The buffers are Array type container objects that carry SDR data, and some carry other things.

The NetworkAPI uses this 'Spec' to choreograph the flow of data between regons as defined by the Links that have been created when setting up a NetworkAPI application. The Links have some properties:
- **Types in API:** The data passed around between regions can be just about anything but the API has required that it be primary data types to simplify the Python interface. So if we pass the new SDR object we need solid .py bindings to define how they are passed, and continue to support the primary data types for older .py code so we don't break the API. 
- **Types in Spec:** The data type of a buffer (as defined in the Spec) on the input of a link must match the data type of the buffer on the other end; no conversion. Consequently, almost all Specs use Real32 as the type. I would like to add conversion if the types do not match so there is a little more flexibility. This would not break the API because existing apps would still work.
- **One-to-one operation** is for a link to connect from one region's output to another region's input. Very simple. The sending region writes to the buffer supplied to it by the Link and the receiving region reads directly from the same buffer that is passed to it by the Link.  The output buffer can be passed without copy.
- **Fan-out operation** is when many links from one output connect to many different inputs, each input gets the same data. Again, the output buffer can be passed without copy. Since the Array uses a shared_ptr each input is actually another instance of the shared_ptr and the actual buffer is deleted only when all instances go out of scope.
- **Fan-In operation** is when links from many outputs connect to a single input. To that region, the input buffer is wide enough to hold all data from all incoming links. The Link object takes care of copying the data to the correct part of the input buffer.  There are some variations on this; the inputs can be OR'd or AND'd together rather than appended to the width.
- **Scatter/Gatherer maps:** This is rather complicated and I have never seen this actually used. The Output bits can be assigned to individual bits of the Input.  This is driven by the LinkPolicy.
- **Delayed Propagation:** Normally data is sent in-sync with execution. All regions are executed and output buffers are moved over the connected links at the end of the execution cycle. But sometimes you want to delay the delivery by a few execution cycles. The Link class implements a delay queue of Array objects so that the data is delivered at the right times.
- **External buffers:** sometimes a region is implemented by Python. Its arrays are numpy structures. To avoid copying the buffer every time it crosses the program language barrier, we can capture the buffer used by numpy as an output, encapulated it into the Array container and pass it around to the C++ implemented regions as input. The buffers would not need to be copied.

Internally the ArrayBuffer object contains a std::shared_ptr for the data and fields for array size and NTA_BasicType type. The buffers are dynamically allocated when the data is assignd. There is a special deallocator so that it handles arrays properly. In the case of numpy arrays there is a special deallocator that does nothing since it should not delete the array when it goes out of scope.

In a region implementation, it uses the array from the input buffer provided by the Link class to call the execute( ) function of the algorithm. It also would pass in a pointer to the Link provided output buffer so that the algorithm can write directly into the region's output buffer without a copy and then the Link passes that buffer pointer to the input of the next region ... all without a copy unless the link requires special handling.  With the new SDR class, we loose the ability to pass data around without copy while going into and out of the Link if the Link only supports flat arrays. However, if both sides need an SDR and the Link knows about SDRs, we can just pass the SDR from the output to the input without a copy.

**Comment by dkeeney** [comment]
Here is an outline of what I would like to do with the Array class.
This is a strawman so what do you think would make better classes?

**new type declarations:**   
```
  nupic/types/Types.hpp
    NTA_BasicType_SDR,     This means that the payload of the Array is a shared_ptr 
                                           to an SDR object.
    NTA_BasicType_Sparse,  This means that the payload of the Array is a UInt array 
                                           of indexes of 1 values. The Array object does not need 
                                           anything special to handle this type. Just treats it
                                           as UInt.
```                           
**Implementation:**
	The SDR is dynamically allocated and placed in a shared_ptr.
        The type will be NTA_BasicType_SDR.


**ArrayBase operations for SDR:**
```
  ArrayBase .hpp
    ArrayBase (NTA_BasicType_SDR)  - modified constructor, creates an Array with an 
                                                   empty SDR as payload.
    ArrayBase (NTA_BasicType_Sparse)  - modified constructor, creates an Array with an 
                                                   empty Sparse array as payload.
    ArrayBase (SDR *s)   - new constuctor, creates an Array containing an existing SDR as
                                                   the payload.  No buffer copy.
    a.allocateBuffer(dim1, dim2, ...);  // will set the size of the buffer in the SDR.
    a.zeroBuffer();                     // will zero out the SDR
    a.releaseBuffer();                  // Deallocates the SDR.  Array is empty.
    a.setBuffer(SDR* s);                // Will include the SDR into the Array but will not delete 
                                      // it when it goes  out of scope (primarily for numpy arrays).
    SDR* s = (SDR*)a.getBuffer();   // returns a pointer to the SDR payload.  You can do 
                                     // anything to the SDR that is inside the Array
                                    // NOTE: this would break some code because most are 
                                   // expecting a pointer to contiguous array. Caller must 
                                    // check type to know how to cast it.
    a.getCount();                   // returns the flattened size of the data in the SDR
    a.getType();                    // returns NTA_BasicType_SDR
    a.save(stream)
    a.load(stream)
    a.operator>>
    a.operator<<
    
    a.getMaxElementsCount();        // related to capacity...throws not implemented  for an SDR.
    a.getBufferSize();              // size in bytes...throws not implemented for SDR.
    a.setCount(len);                // Not implemented for SDR
```    

**Array operations for SDR:**    
```
    a.fromVector(std::vector<T>& v);  // can be used to allocate and populate the 
                               // SDR from a vector. 'a' previously allocated as an SDR.
    
    Array a(NTA_BasicType_SDR, Byte* buffer_ptr, length)     // allocate and populate the internal 
                               // SDR from an array with conversions. with buffer copy.
    Array a(NTA_BasicType_SDR, UInt* buffer_ptr, length)
    Array a(NTA_BasicType_SDR, Real* buffer_ptr, length) 
            ... etc.
    
    Array b = a;           // assuming 'a' contains an SDR, this will make another instance 
                       // of the Array containing the same SDR. Operations on 'a' are 
                      // seen by 'b'.  Operations on 'a' are seen by 'b'.
                      // This is a shallow copy so the data is not copied.  Both instances 
                      // must go out of scope before the SDR is destroyed.
    Array(Array& a)     //  copy constructor makes shallow copy of Array so no buffer copy. 
                    // Same as an assignment.
                           
    Array b = a.copy();    // This makes a deep copy. After the assignment both are independent.
    
    Array b = a.subset(offset, count);    // deep copy of specific portion of array
    
    vector<T> c = a.asVector();    // obtains a vector from contents of the SDR.  
                                      // Maybe optional index argument so it can fetch a row.
    
    a.fromVector(vector<T>& v)  // populates the internal SDR with the vector contents 
                                     //with conversions. (type of Array previously set to SDR)
    
    Array b = a.get_as(type);   // converts data in a to specified type. Full copy
```

**Comment by Thanh-Binh** [comment]
@dkeeney: I find your proposal very interesting and super good. It allows a smart data transfer between regions. 
In the current version dense data are sent so that every time we have to form dense data from sparse for outputing and convert them back to sparse after receiving at receiving regions. By handling the mismatching of data types (region in-/outputs or Specs etc.) I wish me to have any check providing something’s like error messages.

**Comment by dkeeney** [comment]
> The conversion from SDR-> BasicType_Sparse would also be without copy, right? 

As I understand it, the SDR class is implemented with three different formats as vectors. If the SDR's last use was flat sparse then yes, it is possible to grab the raw buffer without a copy.  If not, then it will require a conversion which is equivalent to a copy.  But even if we could grab the raw buffer it is not that simple to pass around. We would have to guarantee that the SDR remained in scope as long as the Array existed because the SDR's vector would retain ownership of the allocated buffer. Once an Array reaches its destination region, how would you turn that back into an SDR when a different vector owns the allocated space? So there would have to be at least one copy.

So to minimize the number of times we have to copy the data (or convert it) we would have to pass a pointer to the entire SDR object.  Note also that the original SDR object must be dynamically allocated into the Array object's shared_ptr to avoid risks of strange behavior.  We must never allow a pointer to be assigned to any more than one shared_ptr. Only one can be the owner.

>  this does add a lot of code, work.

Adding support for SDR's in the Array object does require touching a lot of things but only one of the functions is new (the constructor with an *SDR as input).  I just need to 'adjust' each of the existing functions to do the right thing with a class object rather than an array.  I don't think there is that much code involved.

**Comment by dkeeney** [comment]
Oh, one more thing.  If we are using the Array class to transport the SDR objects between regions via the links, the Link class must be the one that allocates the Array and the SDR objects and presents them to the regions.  The region implementations pass the SDRs to the algorithms.  The algorithms do not allocate the SDR's used for input and output.  Only the Link knows when the Array payload needs to be converted or copied and when it can be passed without a copy as it moves between two regions.

**Comment by ctrl-z-9000-times** [comment]
This looks like a good plan for integrating SDR's into the engine framework.  This does look like a lot of code & complexity...

Question: for a region to send or receive an SDR, what modifications need to be made to the region?  Does it need to have a compute method overload for SDR's?

I'm not sure what value this would add, to the engine framework.  I think that the engine and SDRs solve similar problems: how to orchestrate algorithms and connect the data flows between them.

**Comment by dkeeney** [comment]
>  I think that the engine and SDRs solve similar problems: how to orchestrate algorithms and connect the data flows between them.

Not at all the same.  Arrays and SDRs have similar objectives however ... to act as a container for data but the Arrays are for any data not just SDR data. So it seems logical to allow an SDR to be encapsulated inside an Array object.

**Comment by dkeeney** [comment]
I already have a problem with my proposed design...
```
    ArrayBase (NTA_BasicType_SDR)   - new constuctor, creates an Array with an empty SDR as payload.
```
The problem is that the SDR needs dimensions at creation time.  I do not see any facility in the SDR class for setting or changing the SDR after it is created.

The Link objects in the NetworkAPI that would be creating the buffers do know about dimensions (see the Dimension object) but I don't think this is the same set of dimensions that one might want to use in an algorithm. The Dimension object are intended for use in the Splitter/Gatherer facility in LinkPolicy.  

I do not know the algorithms well enough to know if all algorithms would expect to use the same dimensions in an SDR.  For example, would the SP module expect to see an SDR with the same dimensions that a TM module would expect.  Who or what decides what the dimensions should be? Is it always fixed?

Just thinking off the top of my head, would it be reasonable to have a function in SDR something like [numpy reshape( )](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.reshape.html) ?
The underlining array is always a flat dense array.  When the reshape( ) function is called it sets the dimensions which are used to remap indexes into the flat array to form a multi-dimensional projection of the array. I think this is how it is done in the SDR at( ) function.   Of course the sparse array would be invalidated so it would have to be re-built the next time it is accessed.

I don't know.  @ctrl-z-9000-times  Am I asking for too much?

**Comment by ctrl-z-9000-times** [comment]
> OPF

How many ways are there to make an SP & TM?
1. C++ algorithms
2. C++ engine
3. Py algorithms *+
4. Py bindings + <- IMO the *best* one
5. Py engine *+
6. Py OPF *+

\* From the Numenta repo
\+ Not tested, bindings are partially tested

> I do not see any facility in the SDR class for setting or changing the SDR after it is created.

I can add a reshape method to the SDR.

**Comment by dkeeney** [comment]
> How many ways are there to make an SP & TM?

There are only two.  C++ algorithms and Py algorithms.  Py bindings are just language syntax conversion and should contain no logic.  The Py engine is just a wrapper for the C++ engine so it can be called from Python (also no logic).   Everything else is a hierarchy of abstraction.

Hierarchy:
1. OPF - allows apps to be built without code - all from parameter tables.  Might have a GUI too. Calls engine
2. C++ engine  -- orchestration of a lot of algorithms, calls algorithms
3. C++ and Py algorithms  -- the real thing,  Duplicates for good reasons.

All of this is important.

It is possible for anyone to write their own experimental applications using any level of the hierarchy depending on how much abstraction they need.

In addition there is numenta's htmresearch with its own C++ htmresearch-core but they rely on the normal numenta C++ core (both engine and algorithms) for the basic stuff.  We want to be good enough that we can replace the numenta C++ core when Python2.7 becomes end-of-life.

**Comment by dkeeney** [comment]
@ctrl-z-9000-times I have a situation that I am not sure how to deal with.  This is the Fan-in condition.  Given an input buffer (an SDR) and I have two or more outputs connected to it.  

With raw arrays I just allocate a buffer large enough to hold all of the arrays in parallel and copy each of them in turn with an offset to the beginning of their portion.

However, with an SDR what would an algorithm expect to see?  Should I add another dimension equal to the number of incoming links or should I just increase the size of the topmost dimension to be the sum of the widths of all incoming links?  I would expect that latter but thought I should check.

I do not see a way to set a range of values into an SDR with some sort of iterator.  Start at some index and copy n values.  I could copy in the data one byte at a time, or I could build the composite buffer first and then pass it all into the SDR. How would you recommend doing this?

**Comment by ctrl-z-9000-times** [comment]
@dkeeney ,  I would recommend writing directly to the dense array:
```
SDR output = SDR( room-for-both-regions-to-write-to )
auto &buffer = output.getDense() <- this ref (&) is important, otherwise this will copy!
region1 writes output to buffer[ 0 - half ]
region2 writes output to buffer[ half - end ]
output.setDense( buffer )
```
The last line: Since the buffer is the correct data type, this will not copy the buffer, just swaps it in and does the book-keeping.  Since the buffer was originally allocated by this SDR, it actually swaps vector data with itself so no data is allocated/freed by this line.

**Comment by ctrl-z-9000-times** [comment]
> Should I add another dimension equal to the number of incoming links or should I just increase the size of the topmost dimension to be the sum of the widths of all incoming links? I would expect that latter but thought I should check.

I don't know.  The one big constraint on dimensions is that SDR  must be rectangular, so you might need to flatten everything and concatenate 1-D vectors if all of the inputs are different sizes/shapes.

**Comment by dkeeney** [comment]
> SDR must be rectangular

Hmmm, I guess that makes since. For Fan-in the best I can do is build a 1-D SDR.

**Comment by dkeeney** [comment]
@breznak  I am working on your favorite NTA_BasicType_Sparse within ArrayBase.  Conversion from a dense to sparse is fairly straight forward.  The size in the Array is the number of on bits.

However, converting the other way, from sparse to dense, I have no information to tell me how large the resulting dense buffer should be.  At the time we create the sparse matrix we have that info so I was thinking about some way to carry that information in the buffer.  Maybe use the first element or something but that would probably mess up something.  

I would also need that information for merging multiple buffers in a fan-in situation. Need to offset the indexes by the buffer offset distance.  Any ideas?

**Comment by dkeeney** [comment]
Given that we have the SDR object that has already solved all of those problems; do we still need this NTA_BasicType_Sparse data type?

**Comment by ctrl-z-9000-times** [comment]
> For Fan-in the best I can do is build a 1-D SDR.

I have some ideas about how do fan-in and fan-out operations and -no surprise- they involve subclasses of SDRs.  We could make something like:
```
class SDR_Concatenate( input-SDRs, axis ) public SDR
```
* Fan-In: The new class would register callbacks with input-SDRs so that it knows when it's data is out of data and to update all of the data from that sdr.  

* Fan-Out is done by assignment to the SDR-Concatenation.

* This class would keep all of the normal SDR stuff, including its own data buffers, so it copies the data, but only when it needs to, and only what it needs to.

* And because its a data copy, you can concatenate along any axis.  Concatenating along an axis which isn't the last axis involves writing rows of data, and the rows are not contiguous.  So writing directly to the final location is problematic because the algorithms don't expect a 2D non-contiguous array.  

* Everything still needs to be rectangular.

**Comment by dkeeney** [comment]
Hmmm, sounds interesting.  I hacked up something in ArrayBase to handle it but it is messy and does not handle the dimensions at all.  Let me clean up what I have a little more and then I will push it up to the repository so you can see where it might fit in.

**Comment by dkeeney** [comment]
See push #234

**Comment by dkeeney** [comment]
I don't think that Fan-Out is a problem because the SDR's are contained in a shared_ptr and I can present the same output to multiple Region Inputs.  They will all see the same SDR instance.  However, if we expect algorithms to modify their input SDR's then we will have to make copies of them on any attempt to change something.

**Comment by dkeeney** [comment]
Not sure I understand what you say for Fan-In.  The data will come from multiple region outputs but  all outputs will be concatenated into the input SDR before any region gets executed.  So the regions..and thus the algorithms only see complete data sets.

**Comment by dkeeney** [comment]
> Everything still needs to be rectangular.

I think I understand why that may be needed but how does one determine those values.  The API includes several ways for dimensions to be set but all require a person entering some numbers.  

Also, keep in mind that most regions currently have dimensions turned off (flat array only).  SP and TP in particular in both python and my C++ port. In fact, Of the built-in C++ regions, TestNode (which does not count) and VectorFileSensor allow dimensions to be set.
```
  ns->singleNodeOnly = true;
```
This value in the spec of the region impl says flat files only.  I think you will also find this setting on the .py regions as well.  So if you are requiring dimensions for any type of performance then perhaps we need to think about how we might come up with an algorithm which computes optimum dimensions given the flat array size.

**Comment by dkeeney** [comment]
It could be that dimensions were an 'extra' capability available to those that spent the time to code directly to the algorithms.  There is a LOT of code in the Links and LinkPolicy to support dimensions but for some reason they were turned off.  Maybe they did not really work or maybe the added benifit was not enough to bother with...I don't know.

At one point about a year ago I wrote a whole document describing how the Links worked both in C++ and .py.  I presented it to numenta.  They said they would add it to the documentation.  They may have but I have not seen it anywhere and don't know where to look.  The purpose of that exercise was to ask if dimensions in the Links was code rot and should be removed since it was not really being used.   I got a confirmation that it was, but that confirmation was not very enthusiastic so I suspect nobody really knows.

**Comment by ctrl-z-9000-times** [comment]
> I got a confirmation that it was [code rot], but that confirmation was not very enthusiastic so I suspect nobody really knows.

That might have been embarrassment, for the one who wrote it.  If they don't have good public documentation about how the LinkPolicy works, then probably no one is using it for lack of knowing how.  If the creators have confirmed that it's code rot then my vote is to remove it.

**Comment by ctrl-z-9000-times** [comment]
> Not sure I understand what you say for Fan-In.

Sorry that was a confusing explanation.  The class should make the following code example work:
```
SDR A
SDR B
SDR_Concatenator C( A, B )
A.setDense( data )
B.setDense( data )
C.getDense() -> A & B concatenated
```
And `C.getDense()` will always the latest data.

This class does not do Fan-Out, my mistake.  It could do a split operation. The split operation is the reverse of the fan-in operation. Continuing the example:
```
C.setDense( data )
A.getDense() -> some-of-the-data
B.getDense() -> remainder-of-the-data
```

**Comment by dkeeney** [comment]
I don't know if we need a split operation... although after a fan-In, the algorithm might need a split in order to work with the resulting data. My understanding of the objective of the Fan-In is to provide the algorithm with a larger data set after a fan-in. But there are other things that could be done like OR the data sets together.

You and I are accustomed to working with production code where there is a clear design and performance goals. It is clearly spelled out by marketing what features are needed by the customers. That is not the case here. This code was evolved by experimentation for experimentation.  So in some places it is clearly over-engineered with lots of features that someone thought might be useful in dreaming up new ideas.  The Link object is one of those I think.  

On the other hand there are parts of the code that are not really complete, poorly implemented interfaces, inefficient implementation, and perhaps even wrong and no unit tests.   There are also portions of the code that were early versions of a feature that were subsequently replaced but nobody wanted to remove the old feature because somebody might still be using it.  The BacktrackingTM comes to mind, although I am not sure if it is obsolete or not.

I see our job is to clean up the code as best we can such that the result has the same functionality but more stable. I don't think we will have anyone saying this feature is useful and that one is not.

The new TBT theory and grid cells and the hex grid may make everything in this library obsolete although they say it is built on top of HTM. But it will highlight the features that are most useful, perhaps some that are no longer useful, and some that need to be added.

**Comment by ctrl-z-9000-times** [comment]
> The new TBT theory and grid cells and the hex grid may make everything in this library obsolete although they say it is built on top of HTM. 

Their new "TBT" is about the larger scale organization of the brain, it does not discuss the processing that happens within each macro-column.  It describes some of the output requirements for the macro-columns, which Numenta deduced from understanding the problems that the cortex solves.  In short: every cortical area outputs a representation of what it's seeing (L2/3 cells) in its feed-forward proximal input, and also what its input is doing (motion, L5TT cells).

Numenta doesn't have an implementation of grid cells, except as an encoder which accepts a 2-D coordinate.  I have some ideas about how to make grid cells, as a modified SP, see (Kropff & Treves 2008) for details.  I've prototyped it and it works as advertised, although not quite as well as the source paper since I simplified my model a lot.

However the "Column Pooler" theory is going to make this repo obsolete if we don't port it at some point.  While the SP & TM model cells in cortical layer 4, the column pooler models cortical layers 2/3, and probably also layer 5.

**Comment by dkeeney** [comment]
One of the existing 'features' that is a real problem is Serialization.  There have been at least three attempts and coming up with a reasonable way of making an HTM app's learning persistent over time.  I think we still do not have it right with the save( ) and load( ) functions.

The brain does not do it that way. There is no serialization and HTM is all about trying to do things with a biological foundation. I got to thinking that what our system needs to replace serialization is some sort of always present persistent storage.  

A memory backed file came to mind.  When the HTM learns something there are key data structures that represent the state. Synapses and connections and .. I don't know maybe some other stuff.  What if we could condense down just the bare minimum of that data that represents that learned value and allocate memory for that from space that is backed with a file.  If the program should stop, the learned values persist.  The program could be restarted with its set of parameters, re-initailize all of the transitive structures and re-attach to the persistent learned values.

It may be too early in the research for such a feature but it is an interesting idea.

**Comment by ctrl-z-9000-times** [comment]
> a real problem is Serialization.

IMO the problems is that we try to save the serialized-string to file.  Nupic should return a string and tell the user to deal with their OS.   Nupic should not talk to the OS unless it really needs to.  About half of the OS directory is only needed to save/load HTMs from file.  It's a cross platform headache which python can deal with automatically, and the users are going to need to deal with anyways (to specify the file location).

> A memory backed file came to mind.   If the program should stop, the learned values persist.

I'm assuming you mean "If the program should stop *in an uncontrolled way*"

IMO very difficult.  Imagine if the HTM is in the middle of modifying its memory and the computer shuts down or the USB stick containing the memmapped file is pulled out...  The HTM would need to use atomic read/writes to modify its own memory.  How many program save/load properly when their processes are randomly killed?  Text editors do this by pro-actively saving backups, which the current serialization can do.

**Comment by ctrl-z-9000-times** [comment]
> HTM is all about trying to do things with a biological foundation.

About this: the BacktrackingTM has no biological foundation.  Numenta has acknowledged this fact.  However, it performs better than the vanilla TM on Numenta Anomaly Benchmark (NAB).  Maybe someday it won't be state-of-the-art and then we can easily discard it.

**Comment by ctrl-z-9000-times** [comment]
> I don't know if we need a split operation... 

I think you're right we probably don't.

> But there are other things that could be done like OR the data sets together.

SDR's could do this using the same strategy as SDR-Concatenation, but I would give it its own class instead of rolling it in with concatenation code.

**Comment by dkeeney** [comment]
@rhyolight About a year ago I researched the Link features in the NetworkAPI and wrote a document describing how to configure a link.  I submitted it to numenta and they said they would add it to the documentation.  I am doing some more work on the Link in nupic.cpp and thought it would be good to look through that document to refresh my mind.   I did not keep a copy. Do you happen to know where that went?  If not it is ok.

Thanks.

**Comment by breznak** [comment]
@dkeeney the discussion got a bit long, but `NTA_BasicType_SDR` is now supported by the NetworkAPI, 
can we close this?

**Comment by dkeeney** [comment]
Yes, we are well beyond this now.  Closing.
